### ISSUE TITLE
Allow compiling code to a script

### ISSUE THREAD

**Issue by hgl** [babel/babel#1049]
```js
{
	let a = 1;
}
```

=>

```js
{
	var a = 1;
}
```
Since it doesn't use module syntax, it's not wrapped in an IIFE, and including this code in a script tag cause a global `a` variable being defined.

IMO, if the code isn't a module, let declarations in a block should be wrapped in an IIFE.

**Comment by sebmck** [comment]
Yeah this is expected behaviour. It's assumed that you're already using a module loader which will already wrap it in an IIFE to prevent leaks.

**Comment by hgl** [comment]
But what about simply using the code as a vanilla script? 

For example, polyfills don't need a module loader:

```js
{
   let helper = () = {};
   let fetch = () => { helper(); };
   window.fetch = fetch;
}
```
An have babel to just compile it into es5 and include it via a script tag.

**Comment by sebmck** [comment]
Babel just assumes that all code is module code that you pass it and delegates the module bundling and wrapping to the user. I'll probably include an optional `wrap` transformer or something that produces the behaviour you describe as it's been requested a couple times before.

**Comment by hgl** [comment]
I used to believe all es2015 code should be module code. But then [discussion with caridy](https://mail.mozilla.org/pipermail/es-discuss/2015-March/042090.html) makes me to think otherwise.

**Comment by sebmck** [comment]
In ES6 there's scripts and modules. Modules are implicitly strict and allow the use of import/export syntax. Scripts are basically just `<script src="index.js">` and modules are something that you'd like via `System.import("index.js");` or something like that, AFAIK the delivery mechanism for the browser hasn't been 100% decided yet, could be via on attribute on `<script>` or both etc. 90% of users use a module bundler or loader so it's a pretty fair decision. There are a bunch of technicalities in ES6 scripts regarding strict mode (eg. class bodies are strict mode) so it's far easier to just assume module code.

**Comment by hgl** [comment]
Ah, I see what you mean.

I was initially under the impression that any es2015 code not using import/export is a script, which I now realize is wrong. The environment decides it's a module or a script.

I think my issue should be changed to "Feature request for compiling code to a script".  :)

Probably is the `wrap` the transformer you mentioned. I would expect the CLI to offer a "script" module formatter. Does that sound reasonable?

**Comment by sebmck** [comment]
> I would expect the CLI to offer a "script" module formatter. Does that sound reasonable?

Sure, that sounds like a much better solution. Feel free to reopen and change the issue title.

**Comment by hgl** [comment]
Changed the title, but don't have the permission to reopen the issue.

**Comment by hgl** [comment]
When compiling multiple files using the script formatter, would it be better if the helper code is hoisted to top of the output instead of being repeated in each file?

**Comment by sebmck** [comment]
@hgl Not sure what you mean. Are you referring to compiling multiple files with `$ babel`?

**Comment by hgl** [comment]
Yes.

```
$ babel -m script file1.js file2.js
```

If both `file1.js` and `file2.js` contain class definitions, could there only be one `_createClass` and `_classCallCheck` definition in the ouput?

**Comment by hgl** [comment]
Inspired by this, could babel API supports accepting multiple files at once:

```js
var results = babel.transform([{
  code: "// code string",
  options: { filename: "/foo.js" } // options for each file, overrides default options.
}], options /* default options for all file */);

results[0].code // compiled code of the first file, excluding helper code.
result.helpers // helpers code needed for all files.
```

And a module loader system, especially went bundling all files, concatenates each code, and adds that helpers code at an outer scope. The advantage over external helpers is that:

- Smaller helpers code, only what is truly needed is added.
- No global variable leakage like external helpers does.

@guybedford could systemjs builder benefit from this?

**Comment by sebmck** [comment]
> No global variable leakage like external helpers does.

See the [runtime](https://babeljs.io/docs/usage/runtime/).

**Comment by hgl** [comment]
Oh, do you mean importing the runtime as a module, didn't know it's possible. I guess I should carefully go through the babel doc before making more suggestions. :)

So the only benefit is smaller runtime code. Not sure if you want to pursue this.

**Comment by hgl** [comment]
Just realized importing the runtime as a module probably won't work for scripts, since they won't use cjs or amd formats, so there is no `require` or `define` support.

**Comment by sebmck** [comment]
Actually I don't think this belongs in Babel, at least not at this time. Seems like a slippery slope to support not only multiple file compilation (something partially related is coming) but script support. It's not as easy to just wrap the file in a closure and hoist variables. As I said previously there's implicit strict mode inside classes and this is the first time this has been requested so the demand is very low for the effort and maintenance requirements.

**Comment by hgl** [comment]
What if only strict scripts are supported? I doubt people using babel would explicitly need non-strict code. If scripts are strict, there is no need for considering implicit strict mode, all code is in strict mode.

The use case for script format are mostly for writing polyfill code in es 2015.

I can do some investigation on the code base to see if I can manage to send a PR, would you be interesting in merging a strict mode only script formatter?

**Comment by sebmck** [comment]
> I doubt people using babel would explicitly need non-strict code. If scripts are strict, there is no need for considering implicit strict mode, all code is in strict mode.

Well that's exactly what a script is. No implicit strict mode and no modules.

> The use case for script format are mostly for writing polyfill code in es 2015.

As was mentioned in the esdiscuss thread you linked, just because it's not recommended doesn't mean you can't do it. Polyfills are hacks and violate best practices in the first place (mutating globals).

> would you be interesting in merging a strict mode only script formatter?

Leaning more towards no because the use-case is so limited. There's already the ability to provide [custom module formatters](http://babeljs.io/docs/usage/modules/#custom).

**Comment by fkling** [comment]
> Polyfills are hacks and violate best practices in the first place (mutating globals).

I don't want to derail this discussion, but this is a weak argument against polyfills IMO. The advantages (namely using next gen's API today) outweighs this.

**Comment by sebmck** [comment]
> I don't want to derail this discussion, but this is a weak argument against polyfills IMO. 

It's not an argument against polyfills, I think they're prefectly acceptable and necessary. Arguing for Babel to support scripts just so satisfy the largest use-case that's apparently polyfills (not sure I agree with this) is unnecessary.

**Comment by fkling** [comment]
@sebmck: Ok, then I just misunderstood your statement. I agree.

**Comment by hgl** [comment]
> Well that's exactly what a script is. No implicit strict mode and no modules

The real advantage a script over a module is that it has reliable way to access the global object (e.g., variables declared in the root scope add to the global object). I doubt people look for the no implicit strict mode feature when they use a script. Assuming strict mode for a script does deviate from the spec, but it shouldn't cause trouble for developers. When writing new code, not using strict mode is a wrong thing to do in the first place.

> As was mentioned in the esdiscuss thread you linked ...

To be honest, I'm not absolutely sure what he meant by that, to me it looks like "yeah, you can write polyfill in module if you must, but there is a better way, write the polyfill as a script." and I think it makes sense. Getting a hold of the global object in a module is not reliable (especially you are writing polyfill for the language's stdlib, you really shouldn't rely on `window` or  `global`, not to mention the hack es6-shim uses).

> There's already the ability to provide custom module formatters

Great. The only things missing are allowing assigning to `this` and hosting runtime code to top when there are multiple files.

---

> Arguing for Babel to support scripts just so satisfy the largest use-case that's apparently polyfills (not sure I agree with this) is unnecessary.

I believe the concern mostly raises from the actually implementation complexity? Is the runtime hoisting the most complex one? Wrapping each file's blocks with `(function() { ... }).call(this)` and allowing `this.a = 1` doesn't seem to be that challenging (maybe I'm too naive on this?).

Or you think non-strict script is a must?

**Comment by sebmck** [comment]
> The only things missing are allowing assigning to `this` 

[Disable the strict transformer](http://babeljs.io/docs/usage/transformers/other/strict/#blacklisting).

> and hosting runtime code to top when there are multiple files.

Not a use-case that'll be supported since as you can already assume, it's assumed that you'll be using a module loader since Babel doesn't deal with module bundling at all.

> To be honest, I'm not absolutely sure what he meant by that, to me it looks like "yeah, you can write polyfill in module if you must, but there is a better way, write the polyfill as a script." and I think it makes sense.

Basically. As I mentioned before you can [disable the strict transformer](http://babeljs.io/docs/usage/transformers/other/strict/#blacklisting) to get top-level access to `this`.

> I believe the concern mostly raises from the actually implementation complexity? 

No it's easy. I just don't see the point in including a feature that will see so little use. It has varying levels of complexity depending on the approach that's taken.

> Or you think non-strict script is a must?

Yes. An implictly strict script doesn't make any sense since then it's basically a module. And if it's not implicit then what does it do if there's no strict directive? Throw an error? The behaviour isn't really obvious and this feature would see such little use that it's not worth it.

**Comment by hgl** [comment]
> Disable the strict transformer.

Wow, didn't know it also alleviates assigning to `this`, thanks for the heads up.

Can a custom module formatter wrap root-level blocks with IIFEs or I need to write a custom transformer?

> An implictly strict script doesn't make any sense since then it's basically a module.

You can do this in an implictly strict script, but can't in a module:

```js
if (typeof Symbol == "undefined") {
  Symbol = ... // have the Symbol variable available in the root scope
}
```

> And if it's not implicit then what does it do if there's no strict directive?

Not sure what you mean by "there's no strict directive". By "strict directive" do you mean `"use strict"`? why missing that have to cause an error to be thrown?

> Babel doesn't deal with module bundling at all.

Could Babel at least make bundled code more efficient (size-wise)? This limits to scripts. Currently to bundle the code Babel generates, you basically have two options:

- Repeat the runtime in each file
- Or, include the whole runtime even though you only use a sub-set of it.

By supporting the multiple files API I mentioned eariler, it's possible for a bundler to include only the runtime needed and not repeating it in each file (I mean each section in the final bundled code).

Systemjs builder by default does the repetition, but I think it really should default to this.

**Comment by sebmck** [comment]
@hgl 

> you basically have two choices:
>
> - Repeat the runtime in each file
> - Or, include the whole runtime even though you only use a sub-set of it.

This is incorrect. As I said before there's the [runtime](http://babeljs.io/docs/usage/runtime) and you can [selectively build the external helpers as well as specify a custom format](http://babeljs.io/docs/usage/external-helpers/#selective-builds).

> By supporting the multiple files API I mentioned eariler, it's possible for a bundler to include only the runtime needed and not repeating it in each file (I mean each section in the final bundled code).

If you're using a bundler then you can just enable the `runtime` transformer.

> Systemjs builder by default does the repetition, but I think it really should default to this.

[It wont in the future](https://github.com/systemjs/builder/issues/105).

---

Basically what it boils down to is that there are **tons** of ways to get around the inline helpers. Whether it's using the external helpers or the runtime transformer there's bound to be one that satisfies every use-case.

**Comment by hgl** [comment]
> you can selectively build the external helpers as well as specify a custom format.

Apology for my ignorance. Just tested it, the "var" format is exactly what I'm looking for, and `buildExternalHelpers` is able to handle array with repeated items, really awesome.

I now think it's actually possible to compile to a scripts with the Babel API, by black listing the strict transformer and selectively building the external helpers.

All I need now is this:

```js
{ let a = 1; }
```
->
```js
(function() { var a = 1; })();
```
Am I correct that you are not interested in supporting this? Is it possible with the Babel API ?(possible to inject a custom transformer?)

**Comment by sebmck** [comment]
> Am I correct that you are not interested in supporting this?

Unfortunately.

> (possible to inject a custom transformer?)

It'd be very much possible with the upcoming plugin API (#987). You can already do this with a custom module formatter however:

**node_modules/foo/script.js**

```javascript
export default class ScriptFormatter {
  constructor(file) {
    this.file = file;
  }

  transform(program) {
    var declarators = [];
    var bindings    = this.file.scope.getAllBindingsOfKind("var");
    var body        = [];

    //

    for (var i = 0; i < bindings.length; i++) {
      declarators.push({
        type: "VariableDeclarator"
        id: bindings[i].identifier,
        init: null
      });
    }

    if (declarators.length) {
      body.push({
        type: "VariableDeclaration",
        kind: "var",
        declarations: declarators
      });
    }

    //

    body.push({
      type: "CallExpression",
      callee: {
        type: "FunctionExpression",
        params: [],
        id: null,
        body: {
          type: "BlockStatement",
          body: program.body
        }
      },
      arguments: []
    });

    //

    program.body = body;
  }

  exportDeclaration() {}
  exportAllDeclaration() {}
  importDeclaration() {}
  importSpecifier() {}
  exportSpecifier() {}
}
```

```sh
$ babel --modules script --blacklist strict script.js
```

No idea if that'll work, haven't tested it, probably relies on some APIs in the `experimental` branch.

**Comment by sebmck** [comment]
Whoops, that excludes the blaring emission of transforming the inner variable declarations to assignments.
